### PR TITLE
Redraw all in conhost when changing the palette

### LINKED ISSUE #399: SetConsoleScreenBufferInfoEx color palette behavior broken
19H1 has many improvements concerning console colors, but introduced a new bug. (@zadjii-msft is gonna hate me, sorry)
Testing on 19H1 build 18362.1 and comparing with 1809 build 17763.402.

The console was designed to work like a CGA text screen, with modifiable palette like on VGA.
Whenever the palette is changed, the characters already on screen are displayed using the updated palette on its next refresh cycle,
This was how it used to work until Win10 1809...
The new console code doesn't refresh all displayed cells when SetConsoleScreenBufferInfoEx is called.
This is a breaking compatibility change in 19H1, and an undesirable one as some apps might take advantage of the ability to change the palette without having to redraw the whole text screen to change displayed colors. (like how VGA games used colors cycling).

See the following sample code, it outputs text in color # 7 over a background of color # 0.
Since both lines it outputs are using the same colors indexes, they should be shown in the same color, changing the palette applies to the whole screen, including any existing output.

On 19H1, the console screen isn't updated with the new colors until the user either goes into console settings and clicks OK, or tries to resize the window even just a bit, or even just minimizes and restores the window, forcing a repaint.
This is clearly a bug, as minimizing and restoring the window would not change the colors if it was a new intended behavior.

<pre>
int main()
{
	HANDLE hStdOut = GetStdHandle(STD_OUTPUT_HANDLE);

	CONSOLE_SCREEN_BUFFER_INFOEX csbiex;
	ZeroMemory(&csbiex, sizeof(CONSOLE_SCREEN_BUFFER_INFOEX));
	csbiex.cbSize = sizeof(CONSOLE_SCREEN_BUFFER_INFOEX);
	GetConsoleScreenBufferInfoEx(hStdOut, &csbiex);
	// Make sure we're using color#0 as background and #7 as foreground
	csbiex.wAttributes = FOREGROUND_BLUE | FOREGROUND_GREEN | FOREGROUND_RED;

	// Set blue colors scheme.
	csbiex.ColorTable[0] = RGB(128, 0, 0);
	csbiex.ColorTable[7] = RGB(255, 0, 0);
	SetConsoleScreenBufferInfoEx(hStdOut, &csbiex);
	LPCTSTR pszRedString = TEXT("Bright red on dark red\r\n");
	WriteConsole(hStdOut, pszRedString, _tcslen(pszRedString), NULL, NULL);

	// Set red colors scheme.
	csbiex.ColorTable[0] = RGB(0, 0, 128);
	csbiex.ColorTable[7] = RGB(0, 0, 255);
	SetConsoleScreenBufferInfoEx(hStdOut, &csbiex);
	// At this point, the "Bright red on dark red", as well as all background,
	// for the whole screen should have been turned to blue.
	LPCTSTR pszBlueString = TEXT("Bright blue on dark blue\r\n");
	WriteConsole(hStdOut, pszBlueString, _tcslen(pszBlueString), NULL, NULL);

	getchar();
	return 0;
}
</pre>

Screen on 1809 and below :
1809 and below

Same on 19H1 (18362.1) :
19H1 bug

### PR DESCRIPTION
**Pull request by zadjii-msft**
Does what it says on the tin. Not super complicated of a fix.

* [x] Fixes #399
* [x] I work here.
* [x] Tested manually (not sure there's a great test for this particular path)
gh-399-fixed

### CODE REVIEW

File: src/host/getset.cpp [hunk 1]
```diff
@@ -608,6 +608,18 @@ void ApiRoutines::GetLargestConsoleWindowSizeImpl(const SCREEN_INFORMATION& cont
             gci.SetLegacyColorTableEntry(i, data.ColorTable[i]);
         }
 
+        // GH#399: Trigger a redraw, so that updated colors are repainted, but
+        // only do this if we're not in conpty mode. ConPTY will update the
+        // colors of the palette elsewhere (after TODO GH#10639)
```
Comment (lhecker) [inline src/host/getset.cpp:613]: Where's "elsewhere"?

**Review by lhecker — APPROVED** [review]

File: src/host/getset.cpp (hunk 1 above)
Comment (lhecker) [reply src/host/getset.cpp:613]: Ah, I misread it.

File: src/host/getset.cpp [hunk 2]
```diff
@@ -608,6 +608,18 @@ void ApiRoutines::GetLargestConsoleWindowSizeImpl(const SCREEN_INFORMATION& cont
...
 
+        // GH#399: Trigger a redraw, so that updated colors are repainted, but
+        // only do this if we're not in conpty mode. ConPTY will update the
+        // colors of the palette elsewhere (after TODO GH#10639)
+        if (!gci.IsInVtIoMode())
+        {
+            auto* pRender = ServiceLocator::LocateGlobals().pRender;
+            if (pRender)
+            {
+                pRender->TriggerRedrawAll();
```
Comment (DHowett) [inline src/host/getset.cpp:619]: this shouldn't introduce a poor performance case for applications that aren't intending on changing the colors, right?

**Review by DHowett — APPROVED** [review]

File: src/host/getset.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/host/getset.cpp:619]: > inb4 some crazy app that's calling `SetConsoleScreenBufferInfoEx` in a tight loop

Comment (zadjii-msft) [reply src/host/getset.cpp:619]: I mean, it totally might. I guess I could check the colors to see if they actually changed.

Comment (miniksa) [reply src/host/getset.cpp:619]: > > inb4 some crazy app that's calling `SetConsoleScreenBufferInfoEx` in a tight loop

That's not that crazy. Get/Set console screen buffer on each command is one of the few ways to bulk get/set that sort of state. I wouldn't be surprised if something like PSReadline was doing that without realizing.

Comment (lhecker) [reply src/host/getset.cpp:619]: Every time this is called we `memset` 456 bytes, so it's not entirely cost free.
As far as I can see we don't have any engine anymore which don't already accumulate each row of text during rendering anyways, so we can technically drop any invalidation bitmaps and only use row-wise invalidations. That way the cost of `TriggerRedrawAll()` itself will be negligible. (There's still the cost of redrawing everything of course.)

Apart from that: Why don't we just check whether the palette actually changed here? Let's just cache it here and compare it?

**Review by lhecker — CHANGES_REQUESTED** [review]
Blocking just to be sure to settle our discussion.

File: src/host/getset.cpp (hunk 2 above)
Comment (j4james) [reply src/host/getset.cpp:619]: If you're worried about the performance of `TriggerRedrawAll`, it's worth noting that there's potentially another `TriggerRedrawAll` call a couple of lines down in the `SetDefaultAttributes` method (that only happens if one of the attributes has changed though).

Comment (miniksa) [reply src/host/getset.cpp:619]: If it's not hard to check if they changed... we should do that before forcing a repaint. We've been bitten enough times by apps doing weird things with our APIs.

**Comment by miniksa** [comment]
Detagging so it doesn't disappear on @zadjii-msft since he's a core contributor.

**Comment by zadjii-msft** [comment]
_shhhh bot, it's hack week not bugfix week_

File: src/host/getset.cpp [hunk 3]
```diff
@@ -603,18 +603,27 @@ void ApiRoutines::GetLargestConsoleWindowSizeImpl(const SCREEN_INFORMATION& cont
         }
         const auto newBufferSize = context.GetBufferSize().Dimensions();
 
+        bool changedOneTableEntry = false;
         for (size_t i = 0; i < std::size(data.ColorTable); i++)
         {
-            gci.SetLegacyColorTableEntry(i, data.ColorTable[i]);
+            // Check if we actually changed a palette color
+            const auto& newColor{ data.ColorTable[i] };
+            changedOneTableEntry = changedOneTableEntry || (newColor != gci.GetColorTableEntry(i));
```
Comment (lhecker) [inline src/host/getset.cpp:611]:
```cpp
changedOneTableEntry |= newColor != gci.GetColorTableEntry(i);
```
...works as well.

**Review by lhecker — APPROVED** [review]

**Pull request merged**